transport-classes-io_uring/src/main/java/io/netty/channel/uring/IoUringBufferRingRecvByteBufAllocator.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2025 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.channel.uring;
+
+import io.netty.channel.AdaptiveRecvByteBufAllocator;
+import io.netty.channel.RecvByteBufAllocator;
+import io.netty.util.UncheckedBooleanSupplier;
+
+import java.util.Objects;
+
+/**
+ * {@link RecvByteBufAllocator} that should be used if we want to use provided buffers.
+ * It will return the buffer group id to use when submitting recv / read / readv {@link IoUringIoOps}.
+ * <p>
+ * Check
+ * <a href="https://man7.org/linux/man-pages/man3/io_uring_setup_buf_ring.3.html"> man io_uring_setup_buf_ring</a>
+ * and this <a href="https://lwn.net/Articles/815491/">LWN article</a> for more details.
+ */
+public final class IoUringBufferRingRecvByteBufAllocator implements RecvByteBufAllocator {
+    private final RecvByteBufAllocator allocator;
+    private final IoUringBufferGroupIdChooser idChooser;
+
+    /**
+     * Interface that will return the buffer group id.
+     */
+    public interface IoUringBufferGroupIdChooser {
+        /**
+         * Return the group id that should be used for the given size. This method must return a valid group id
+         * that was configured previously via {
+         * @link IoUringIoHandlerConfig#addBufferRingConfig(IoUringBufferRingConfig)}
+         *
+         * @param sizeGuess the guess of bytes that we think we might be able to read.
+         * @return          the group id.
+         */
+        short choose(int sizeGuess);
```

Comment:
Should the chooser not somehow be informed when io_uring consumes data from a buffer ring group, so it can keep track?

How are buffer rings replenished? The `IoUringIoHandlerConfig` is only processed when the IoUringHander is created, as far as I can see.

Reply:
When the buffer ring is exhausted, we only need to fill the io_uring_buf elements of the io_uring_buf_ring and then advance the tail to dynamically supplement the buffers. 

However, we need to consider how to handle the buffers that exceed the ring size，and the association between these buffer IDs and buffers. Perhaps using a short(buffer id) -> ByteBuf  map  can solve this problem?

 I think if we want to implement this feature, we can call back the chooser when the buffering is exhausted, and the chooser will return whether buffer supplementation is required.

Reply:
I think we only need to know when the ring is exhausted and also when the ring has something spare again. After thinking a bit more about this a user might want to stop reading when the ring has nothing left and so disable auto-read and then start again once there is something to use.

Reply:
> When the buffer ring is exhausted, we only need to fill the io_uring_buf elements of the io_uring_buf_ring and then advance the tail to dynamically supplement the buffers.
> 
> However, we need to consider how to handle the buffers that exceed the ring size，and the association between these buffer IDs and buffers. Perhaps using a short(buffer id) -> ByteBuf map can solve this problem?

Hmm not sure I get what you suggest here... Can you explain a bit more why there needs to be some special mapping ?
> 
> I think if we want to implement this feature, we can call back the chooser when the buffering is exhausted, and the chooser will return whether buffer supplementation is required.

Reply:
I haven’t fully figured out how to support dynamically adding extra buffers yet.
 For now, I believe the current implementation is sufficient，IoUringBufferGroupIdHandler::exhausted is enough for me

Reply:
I wonder if we also need to notify somehow once the ring has something to use again @dreamlike-ocean